File: SWEA/Unrated/1949. ［모의 SW 역량테스트］ 등산로 조성/［모의 SW 역량테스트］ 등산로 조성.java

import java.util.Scanner;

public class Solution {

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		int t = sc.nextInt();
		int tt = 1;

		while (tt <= t) {
			n = sc.nextInt();
			k = sc.nextInt();

			grid = new int[n][n];
			visited = new boolean[n][n];

			int max = 0;
			for (int i = 0; i < n; i++) {
				for (int j = 0; j < n; j++) {
					int tmp = sc.nextInt();
					grid[i][j] = tmp;
					if (tmp > max) {
						max = tmp;
					}
				}
			}

			// 중요한 건 내 위치에서 그 다음 위치를 깎아야 한다면 내 위치보다 -1 만큼만 해주면 되는 것.
			// 깍을 수 있는가?는 1. 아직 안깎았고 2. 현재위치 - k 보다 작아야 함.
			eachmaxlength = 0;
			int ans = 0; // 등산로 길이
			for (int i = 0; i < n; i++) {
				for (int j = 0; j < n; j++) {
					if (grid[i][j] == max) { // 최대 높이인 지점들만 지형 깎는 과정 거쳐서 길이 구하고, 그 중 최댓값이 ans가 될거임
//						ans = Math.max(ans, work(i, j, 1, true));
						ans = Math.max(ans, work(i, j, 1, grid[i][j], true)); // 현재 내 높이 추가
					}
				}
			}
			System.out.println("#" + tt + " " + ans);
			tt++;
		}

	}

	static int n;
	static int k;

	static int row[] = { -1, 1, 0, 0 };
	static int col[] = { 0, 0, 1, -1 };
	static boolean[][] visited;
	static int[][] grid;
	static int eachmaxlength;

	private static int work(int r, int c, int length, boolean onechance) { // 내 위치 i, j 와 거리, 깎을 수 있는지?

		eachmaxlength = Math.max(length, eachmaxlength);

		visited[r][c] = true;

		for (int i = 0; i < 4; i++) {
			int nr = r + row[i];
			int nc = c + col[i];
			if (nr >= 0 && nr < n && nc >= 0 && nc < n && !visited[nr][nc]) {
				if (grid[nr][nc] < grid[r][c]) {
					work(nr, nc, length + 1, onechance);
				} else if (grid[nr][nc] - k < grid[r][c] && onechance) {
					int tmp = grid[nr][nc];
					grid[nr][nc] = grid[r][c] - 1;
					work(nr, nc, length + 1, false);
					grid[nr][nc] = tmp;
				}
			}
		}

		visited[r][c] = false;

		return eachmaxlength;

	}

	private static int work(int r, int c, int length, int height, boolean onechance) { // 내 위치 i, j 와 거리, 깎을 수 있는지?

		eachmaxlength = Math.max(length, eachmaxlength);

		visited[r][c] = true;

		for (int i = 0; i < 4; i++) {
			int nr = r + row[i];
			int nc = c + col[i];
			if (nr >= 0 && nr < n && nc >= 0 && nc < n && !visited[nr][nc]) {
				if (grid[nr][nc] < height) {
					work(nr, nc, length + 1, grid[nr][nc], onechance);
				} else if (grid[nr][nc] - k < height && onechance) {
//					onechance = false; // 이거 안됨!!
					work(nr, nc, length + 1, height - 1, false); // 매개변수를 바꿔버리면 그 뒤에 작업들이 영향을 받고 모두 false 가 되버림!! 안된다!
				}
			}
		}

		visited[r][c] = false;

		return eachmaxlength;

	}

}

File: SWEA/Unrated/1949. ［모의 SW 역량테스트］ 등산로 조성/［모의 SW 역량테스트］ 등산로 조성.py

'''
빠른 탈출 조건은 없음 완탐해야함
벽 부수고 이동하기랑 다른 것.
벽부수기는 벽을 부셨냐 안부셨냐만 따지면 되는데 -> 부서지기만 하면 다른 곳에서도 갈 수 있음
이 문제는 어디서 왔냐에 따라서 높이가 달라지기 때문에 어디서 왔는지의 방문체크도 해야줘야함

bfs로 어거지로 풀어봤는데.....
오랜만에 내 코드를 면밀히 뜯어보는 ...!! 좋으 ㄴ경험ㅇ ㅇ.. 이였습니다..ㅠ...

'''
import copy
from collections import deque

T = int(input())


def bfs():

    while q:
        global ans
        r, c, cnt, bomb, hei, visited = q.popleft()  # 부모 높이를 담아줌
        visited2 = copy.copy(visited) # 각기 다른 루트니까 visited 도 다 다르죠!
        # print(visited,cnt,bomb)
        ans = max(ans, cnt)

        for k in range(4):
            nr = r + row[k]
            nc = c + col[k]
            if (not (0 <= nr < n and 0 <= nc < n)):
                continue
            # 0이면 안부시고 갔음
            # 부셨든 안부셨든 나보다 낮으면 가.
            if (r * n + c, nr * n + nc) not in visited2 and grid[nr][nc] < hei:
                visited2.add((r * n + c, nr * n + nc)) # 두개나 넣어줘야함, 전-> 후, 후-> 전 이렇게 해야 처음 위치로도 안돌아감
                visited2.add((nr * n + nc, r * n + c))
                q.append((nr, nc, cnt + 1, bomb, grid[nr][nc],visited2))

            elif (r * n + c, nr * n + nc) not in visited2 and grid[nr][nc] - kk < hei and bomb:
                visited2.add((r * n + c,nr * n + nc))
                visited2.add((nr * n + nc, r * n + c))
                q.append((nr, nc, cnt + 1, False, hei - 1,visited2))



for tc in range(T):

    n, kk = map(int, input().split())  # 맵 크기와 깎을 수 있는 높이
    grid = [[] for i in range(n)]
    maxH = -1
    for i in range(n):
        tmp = list(map(int, input().split()))
        for j in range(n):
            grid[i] = tmp
            if (grid[i][j] > maxH):
                maxH = grid[i][j]
    ans = 0
    row = [0, 1, -1, 0]
    col = [-1, 0, 0, 1]

    for i in range(n):
        for j in range(n):
            if (grid[i][j] == maxH):
                q = deque()  # q에 바로담기
                v = set() # bfs에 visited 데리고 다닐거임... 완탐이니ㅣ까....
                for k in range(4):
                    nr = i+row[k]
                    nc = j+col[k]
                q.append((i, j, 1, True, maxH, v))
                bfs()
    # print(q)

    print(f"#{tc + 1} {ans}")
